presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownLimitThroughWindow.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.metadata.FunctionId;
+import io.prestosql.metadata.Metadata;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.LimitNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.TopNRowNumberNode;
+import io.prestosql.sql.planner.plan.WindowNode;
+import io.prestosql.sql.tree.QualifiedName;
+
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.SystemSessionProperties.isOptimizeTopNRowNumber;
+import static io.prestosql.sql.planner.iterative.rule.WindowPushdownRules.canOptimizeWindowFunction;
+import static io.prestosql.sql.planner.iterative.rule.WindowPushdownRules.convertToTopNRowNumber;
+import static io.prestosql.sql.planner.plan.ChildReplacer.replaceChildren;
+import static io.prestosql.sql.planner.plan.Patterns.limit;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownLimitThroughWindow
+        implements Rule<LimitNode>
+{
+    private static final Pattern<LimitNode> PATTERN = limit()
+            .matching(WindowPushdownRules::pushdownableLimit)
+            .with(source().matching(window()));
+
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownLimitThroughWindow(Metadata metadata)
+    {
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+    }
+
+    @Override
+    public Pattern<LimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode node, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        PlanNode source = getOnlyElement(lookup.resolveGroup(node.getSource()).collect(Collectors.toList()));
+        int limit = toIntExact(node.getCount());
+        if (source instanceof WindowNode && canOptimizeWindowFunction((WindowNode) source, rowNumberFunctionId) && isOptimizeTopNRowNumber(context.getSession())) {
```

Comment:
Move `isOptimizeTopNRowNumber(context.getSession())` to the `isEnabled()` method (defined in the Rule interface).